ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,508 @@
+        }
+        this.setState({usersLoading: false});
+      });
+  }
+
+  removeCollaborator(user: UserRole): void {
+    this.setState(({userRolesList, workspace}) => (
+      {userRolesList: fp.remove(({email}) => user.email === email, userRolesList),
+        workspace: {...workspace,
+          userRoles: fp.remove(({email}) => {
+            return user.email === email;
+          }, workspace.userRoles)} as Workspace}
+    ));
+  }
+
+  reloadWorkspace(): void {
+    workspacesApi().getWorkspace(this.state.workspace.namespace, this.state.workspace.id)
+      .then((workspaceResponse) => {
+        this.setState({workspace: workspaceResponse.workspace});
+        this.resetModalState();
+      })
+      .catch(({status}) => {
+        if (status === 404) {
+          this.setState({workspaceFound: false});
+        }
+      });
+  }
+
+  addCollaborator(user: User): void {
+    const userRole: UserRole = {givenName: user.givenName, familyName: user.familyName,
+      email: user.email, role: WorkspaceAccessLevel.READER};
+    this.setState(({userRolesList, workspace}) => (
+      {searchTerm: '', autocompleteLoading: false, autocompleteUsers: [], dropDown: false,
+        userRolesList: fp.concat(userRolesList, [userRole]),
+        workspace: {...workspace, userRoles: fp.concat(userRolesList, [userRole])} as Workspace}
+    ));
+  }
+
+  userSearch(value: string): void {
+    this.setState({autocompleteLoading: true, autocompleteUsers: [], searchTerm: value});
+    if (!value.trim()) {
+      this.setState({autocompleteLoading: false, dropDown: false});
+      return;
+    }
+    userApi().user(this.state.searchTerm)
+      .then((response) => {
+        response.users = fp.differenceWith((a, b) => {
+          return a.email === b.email;
+        }, response.users, this.state.userRolesList);
+        this.setState({
+          autocompleteUsers: response.users.splice(0, 4),
+          autocompleteLoading: false,
+          dropDown: true
+        });
+      });
+  }
+
+  setRole = (e, user)  => {
+    const oldUserRoles = this.state.userRolesList;
+    const newUserRoleList = fp.map((u) => {
+      return u.email === user.email ? {...u, role: e.value} : u;
+    }, oldUserRoles);
+    this.setState(({userRolesList, workspace}) => ({userRolesList: newUserRoleList,
+      workspace: {...workspace, userRoles: newUserRoleList} as Workspace}));
+  }
+
+  resetModalState(): void {
+    this.setState({
+      userNotFound: '',
+      workspaceShareError: false,
+      workspaceUpdateConflictError: false,
+      usersLoading: false,
+    });
+  }
+
+  openDropdown(): void {
+    this.setState({dropDown: true});
+  }
+
+  closeDropdown(): void {
+    this.setState({dropDown: false});
+  }
+
+  handleClickOutsideSearch = (e) => {
+    if (this.searchingNode) {
+      if (this.searchingNode.contains(e.target)) {
+        return;
+      }
+      this.closeDropdown();
+    }
+  }
+
+  cleanClassNameForSelect(value: string): string {
```

Comment:
This is just for testing right? What happens if this produces a bad value? Does it result in an error? I'm not sure this regex is bullet-proof for all possible email characters so I want to understand what risk there is here.

Reply:
So not really sure what you mean by "bad value". This is just for testing. So the rest of the time we don't really care what it produces (as long as it doesn't actually break anything). I replaced it with bunch of random things, and it doesn't cause any problems as far as I can tell (tested with values like "k* 0").

Reply:
Thanks - my concern was that it would cause some kind of rendering issue or bug in the application. If not - then I'm not really worried about it.